File: flask-mapping-service/app/utils/pathfinding.py
"""
Pathfinding over the admin-declared node/edge graph.

- Dijkstra with Haversine-computed edge weights (never manually typed).
- Floor-aware full-tour via greedy nearest-neighbor with floor zig-zag avoidance.
- Turn-by-turn instruction generation using geodetic bearings.
"""

import heapq
import math
from app.utils.geo import haversine_distance


# ---------------------------------------------------------------------------
# Dijkstra
# ---------------------------------------------------------------------------

def _build_graph(nodes_dict, edges_list):
    """Build a bidirectional adjacency list from walkable edges.
    Prefers explicit edge distance, then Haversine if GPS present, else Euclidean with scale."""
    graph = {nid: [] for nid in nodes_dict}
    for e in edges_list:
        if not e.get("walkable", True):
            continue
        u, v = e["from_node_id"], e["to_node_id"]
        if u not in graph or v not in graph:
            continue
        nu, nv = nodes_dict[u], nodes_dict[v]
        
        # Priority 1: Explicit stored edge distance
        edge_dist = e.get("distance")
        if edge_dist is not None and float(edge_dist) > 0:
            dist = float(edge_dist)
        # Priority 2: GPS coordinates if available on both nodes
        elif nu.get("latitude") is not None and nv.get("latitude") is not None:
            dist = haversine_distance(
                nu["latitude"], nu["longitude"],
                nv["latitude"], nv["longitude"],
            )
        # Priority 3: Euclidean distance using pixel coordinates and realistic scale
        else:
            x1 = float(nu.get("x_coordinate") or 0)
            x2 = float(nv.get("x_coordinate") or 0)
            y1 = float(nu.get("y_coordinate") or 0)
            y2 = float(nv.get("y_coordinate") or 0)
            dx = (x1 - x2) * 1000.0 if x1 <= 1.0 and x2 <= 1.0 else (x1 - x2)
            dy = (y1 - y2) * 1000.0 if y1 <= 1.0 and y2 <= 1.0 else (y1 - y2)
            # Default scale ~0.035 meters per pixel if not specified
            dist = max(0.5, math.hypot(dx, dy) * 0.035)
            
        edge_type = e.get("edge_type", "normal")
        graph[u].append((v, dist, edge_type))
        graph[v].append((u, dist, edge_type))
    return graph


def find_shortest_path(start_id, end_id, nodes_dict, edges_list):
    """
    Dijkstra shortest path.
    Returns (path_list_of_node_dicts, total_distance_m, instructions_list).
    Returns (None, 0, []) when no route exists.
    """
    graph = _build_graph(nodes_dict, edges_list)

    if start_id not in graph or end_id not in graph:
        return None, 0, []

    dist = {nid: float("inf") for nid in nodes_dict}
    dist[start_id] = 0.0
    prev = {}
    pq = [(0.0, start_id)]

    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        if u == end_id:
            break
        for v, w, _ in graph.get(u, []):
            alt = d + w
            if alt < dist[v]:
                dist[v] = alt
                prev[v] = u
                heapq.heappush(pq, (alt, v))

    if dist[end_id] == float("inf"):
        return None, 0, []

    # Reconstruct path
    path = []
    cur = end_id
    while cur is not None:
        path.append(nodes_dict[cur])
        cur = prev.get(cur)
    path.reverse()

    instructions = generate_instructions(path)
    return path, dist[end_id], instructions


# ---------------------------------------------------------------------------
# Floor-aware full tour (greedy nearest-neighbor, zig-zag avoidance)
# ---------------------------------------------------------------------------

def is_exit_node(node):
    if not node or not node.get("node_type"):
        return False
    tags = [t.strip() for t in str(node["node_type"]).split(",")]
    return "exit" in tags


def plan_full_tour(start_id, exit_id, artifact_ids, nodes_dict, edges_list):
    """
    Greedy nearest-neighbor tour from start through all artifacts to exit.
    Floor-aware: prefers same-floor artifacts before switching floors.
    If exit_id is not specified or invalid, finds the optimal exit door node among all exits.
    """
    if not artifact_ids:
        if exit_id and exit_id in nodes_dict:
            return find_shortest_path(start_id, exit_id, nodes_dict, edges_list)
        else:
            exit_candidates = [nid for nid, n in nodes_dict.items() if is_exit_node(n)]
            best_p, best_d, best_i = None, float("inf"), []
            for eid in exit_candidates:
                p, d, i = find_shortest_path(start_id, eid, nodes_dict, edges_list)
                if p and d < best_d:
                    best_p, best_d, best_i = p, d, i
            return best_p, (best_d if best_p else 0), best_i

    full_path = []
    full_instructions = []
    total_dist = 0.0

    current = start_id
    remaining = list(artifact_ids)

    while remaining:
        current_floor = nodes_dict.get(current, {}).get("floor")

        # Partition into same-floor and cross-floor
        same_floor = [a for a in remaining if nodes_dict.get(a, {}).get("floor") == current_floor]
        candidates = same_floor if same_floor else remaining

        best_id = None
        best_route = None
        best_dist = float("inf")

        for aid in candidates:
            path, d, instr = find_shortest_path(current, aid, nodes_dict, edges_list)
            if path and d < best_dist:
                best_dist = d
                best_id = aid
                best_route = (path, d, instr)

        if best_route is None or best_id is None:
            # Can't reach any remaining artifact; skip them
            break

        path, d, instr = best_route
        if full_path and path:
            full_path.extend(path[1:])
        else:
            full_path.extend(path)
        full_instructions.extend(instr)
        total_dist += d

        current = best_id
        remaining.remove(best_id)

    # Route to exit
    exit_path, exit_d, exit_instr = None, 0, []
    if exit_id and exit_id in nodes_dict:
        exit_path, exit_d, exit_instr = find_shortest_path(current, exit_id, nodes_dict, edges_list)
    
    if not exit_path:
        # Evaluate all exit door nodes and pick the shortest route from current
        exit_candidates = [nid for nid, n in nodes_dict.items() if is_exit_node(n)]
        best_exit_dist = float("inf")
        for eid in exit_candidates:
            epath, edist, einstr = find_shortest_path(current, eid, nodes_dict, edges_list)
            if epath and edist < best_exit_dist:
                best_exit_dist = edist
                exit_path, exit_d, exit_instr = epath, edist, einstr

    if exit_path:
        if full_path:
            full_path.extend(exit_path[1:])
        else:
            full_path.extend(exit_path)
        full_instructions.extend(exit_instr)
        total_dist += exit_d

    return full_path, total_dist, full_instructions


# ---------------------------------------------------------------------------
# Instruction generation
# ---------------------------------------------------------------------------

def _bearing(lat1, lon1, lat2, lon2):
    """Initial bearing in degrees from (lat1,lon1) → (lat2,lon2)."""
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    dl = math.radians(lon2 - lon1)
    x = math.sin(dl) * math.cos(phi2)
    y = math.cos(phi1) * math.sin(phi2) - math.sin(phi1) * math.cos(phi2) * math.cos(dl)
    return (math.degrees(math.atan2(x, y)) + 360) % 360


def _turn_word(angle_diff):
    """Human-readable turn direction from a signed angle difference."""
    d = ((angle_diff + 180) % 360) - 180
    if abs(d) < 25:
        return None  # Straight
    if d > 0:
        return "Turn right" if d < 100 else "Sharp right"
    return "Turn left" if d > -100 else "Sharp left"


def generate_instructions(path):
    """Generate turn-by-turn instructions from a path of node dicts."""
    if not path or len(path) < 2:
        return ["You have arrived."]

    instructions = [f"Start at {path[0].get('name', 'here')}"]

    def _get_dist(n1, n2):
        if n1.get("latitude") is not None and n2.get("latitude") is not None:
            return haversine_distance(n1["latitude"], n1["longitude"], n2["latitude"], n2["longitude"])
        x1 = float(n1.get("x_coordinate") or 0)
        x2 = float(n2.get("x_coordinate") or 0)
        y1 = float(n1.get("y_coordinate") or 0)
        y2 = float(n2.get("y_coordinate") or 0)
        dx_px = (x1 - x2) * 613 if x1 <= 1.0 and x2 <= 1.0 else (x1 - x2)
        dy_px = (y1 - y2) * 370 if y1 <= 1.0 and y2 <= 1.0 else (y1 - y2)
        return max(1.0, math.hypot(dx_px, dy_px) * 0.035)

    def _get_bearing(n1, n2):
        if n1.get("latitude") is not None and n2.get("latitude") is not None:
            return _bearing(n1["latitude"], n1["longitude"], n2["latitude"], n2["longitude"])
        x1 = float(n1.get("x_coordinate") or 0)
        x2 = float(n2.get("x_coordinate") or 0)
        y1 = float(n1.get("y_coordinate") or 0)
        y2 = float(n2.get("y_coordinate") or 0)
        dx = (x2 - x1) * 613 if x1 <= 1.0 and x2 <= 1.0 else (x2 - x1)
        dy = (y2 - y1) * 370 if y1 <= 1.0 and y2 <= 1.0 else (y2 - y1)
        return (math.degrees(math.atan2(dx, -dy)) + 360) % 360

    for i in range(len(path) - 1):
        cur = path[i]
        nxt = path[i + 1]

        # Floor change
        if cur.get("floor") != nxt.get("floor"):
            connector = nxt.get("node_type", "stairs")
            if connector not in ("stairs", "elevator", "escalator"):
                connector = "stairs/elevator"
            instructions.append(f"Take the {connector} to Floor {nxt.get('floor')}")
            continue

        if i == len(path) - 2:
            dist = _get_dist(cur, nxt)
            instructions.append(f"Arrive at {nxt.get('name', 'destination')} ({round(dist)} m ahead)")
            continue

        # Bearing turn detection
        nnxt = path[i + 2]
        if nxt.get("floor") != nnxt.get("floor"):
            dist = _get_dist(cur, nxt)
            instructions.append(f"Walk straight for {round(dist)} m")
            continue

        b1 = _get_bearing(cur, nxt)
        b2 = _get_bearing(nxt, nnxt)
        turn = _turn_word(b2 - b1)

        dist = _get_dist(cur, nxt)
        if turn:
            instructions.append(f"{turn} at {nxt.get('name', 'waypoint')} ({round(dist)} m)")
        else:
            instructions.append(f"Walk straight for {round(dist)} m")

    return instructions


# ---------------------------------------------------------------------------
# Nearby (Haversine-based, graph-agnostic)
# ---------------------------------------------------------------------------

def get_nearby_nodes(start_node_id, radius_meters=None, hops=None, nodes_dict=None, edges_list=None):
    """Return nodes within *hops* graph-hops of start_node_id."""
    if not nodes_dict or not edges_list or start_node_id not in nodes_dict:
        return []
    if hops is None:
        hops = 3

    graph = {n: [] for n in nodes_dict}
    for e in edges_list:
        if not e.get("walkable", True):
            continue
        u, v = str(e["from_node_id"]), str(e["to_node_id"])
        if u in graph and v in graph:
            graph[u].append(v)
            graph[v].append(u)

    visited = {start_node_id}
    queue = [(start_node_id, 0)]
    result = []
    while queue:
        curr, h = queue.pop(0)
        if curr != start_node_id:
            result.append(nodes_dict[curr])
        if h < hops:
            for nbr in graph.get(curr, []):
                if nbr not in visited:
                    visited.add(nbr)
                    queue.append((nbr, h + 1))
    return result

File: mobile/lib/services/pathfinding.dart
import 'dart:collection';
import 'dart:math';

import '../models/navigation_models.dart';
import '../utils/navigation_math.dart';

class RouteResult {
  final List<MapNode> path;
  final double distance;
  final List<String> instructions;
  RouteResult({
    required this.path,
    required this.distance,
    required this.instructions,
  });
}

class PathfindingService {
  final List<MapNode> nodes;
  final List<MapEdge> edges;
  final List<FloorPlan> floorPlans;

  late final Map<String, MapNode> _nodeMap;
  late final Map<String, List<_AdjEntry>> _graph;

  PathfindingService({
    required this.nodes,
    required this.edges,
    required this.floorPlans,
  }) {
    _nodeMap = {for (var n in nodes) n.id: n};
    _graph = {};
    for (var n in nodes) {
      _graph[n.id] = [];
    }
    for (var e in edges) {
      if (!e.walkable) continue;
      final u = _nodeMap[e.fromNodeId], v = _nodeMap[e.toNodeId];
      if (u == null || v == null) continue;
      
      double d = e.distance;
      if (d <= 0) {
        if (u.latitude != null && v.latitude != null && u.longitude != null && v.longitude != null) {
          d = NavigationMath.distanceBetween(
            u.latitude!,
            u.longitude!,
            v.latitude!,
            v.longitude!,
          );
        } else {
          final fp = floorPlans.where((p) => p.id == u.floorPlanId).firstOrNull ?? (floorPlans.isNotEmpty ? floorPlans.first : null);
          final double scale = (fp != null && fp.scaleMetersPerPx > 0) ? fp.scaleMetersPerPx : 0.035;
          final double dx = v.x - u.x;
          final double dy = v.y - u.y;
          d = max(0.5, sqrt(dx * dx + dy * dy) * scale);
        }
      }
      _graph[u.id]?.add(_AdjEntry(v.id, d, e.edgeType));
      _graph[v.id]?.add(_AdjEntry(u.id, d, e.edgeType));
    }
  }

  MapNode? getNode(String id) => _nodeMap[id];

  RouteResult? findShortestPath(String startId, String endId) {
    if (!_graph.containsKey(startId) || !_graph.containsKey(endId)) return null;

    final dist = <String, double>{};
    final prev = <String, String>{};
    for (var n in nodes) {
      dist[n.id] = double.infinity;
    }
    dist[startId] = 0;

    final pq = SplayTreeSet<_PQEntry>((a, b) {
      int c = a.dist.compareTo(b.dist);
      return c != 0 ? c : a.id.compareTo(b.id);
    });
    pq.add(_PQEntry(startId, 0));

    while (pq.isNotEmpty) {
      final cur = pq.first;
      pq.remove(cur);
      if (cur.dist > dist[cur.id]!) continue;
      if (cur.id == endId) break;
      for (var adj in _graph[cur.id] ?? []) {
        final alt = cur.dist + adj.dist;
        if (alt < dist[adj.id]!) {
          dist[adj.id] = alt;
          prev[adj.id] = cur.id;
          pq.add(_PQEntry(adj.id, alt));
        }
      }
    }

    if (dist[endId] == double.infinity) return null;

    List<MapNode> path = [];
    String? cur = endId;
    while (cur != null) {
      final n = _nodeMap[cur];
      if (n != null) path.add(n);
      cur = prev[cur];
    }
    path = path.reversed.toList();

    return RouteResult(
      path: path,
      distance: dist[endId]!,
      instructions: _generateInstructions(path),
    );
  }

  bool _isExitNode(MapNode n) =>
      n.nodeType.toLowerCase().split(',').map((s) => s.trim()).contains('exit');

  /// Floor-aware full tour: greedy nearest-neighbor, avoids zig-zagging floors.
  RouteResult? planFullTour(
    String entranceId,
    String? exitId,
    List<String> artifactIds,
  ) {
    String actualExitId = exitId ?? '';
    if (actualExitId.isEmpty || !_nodeMap.containsKey(actualExitId)) {
      final exitCandidates = nodes
          .where((n) => _isExitNode(n))
          .map((n) => n.id)
          .toList();
      String? bestEid;
      double minExitDist = double.infinity;
      for (var eid in exitCandidates) {
        final r = findShortestPath(entranceId, eid);
        if (r != null && r.distance < minExitDist) {
          minExitDist = r.distance;
          bestEid = eid;
        }
      }
      if (bestEid != null) actualExitId = bestEid;
    }

    if (artifactIds.isEmpty) return findShortestPath(entranceId, actualExitId);

    List<MapNode> fullPath = [];
    List<String> fullInstr = [];
    double totalDist = 0;
    String current = entranceId;
    List<String> remaining = List.from(artifactIds);

    while (remaining.isNotEmpty) {
      final currentFloor = _nodeMap[current]?.floor;
      final sameFloor = remaining
          .where((a) => _nodeMap[a]?.floor == currentFloor)
          .toList();
      final candidates = sameFloor.isNotEmpty ? sameFloor : remaining;

      String? bestId;
      RouteResult? bestRoute;
      double bestDist = double.infinity;

      for (var aid in candidates) {
        final r = findShortestPath(current, aid);
        if (r != null && r.distance < bestDist) {
          bestDist = r.distance;
          bestId = aid;
          bestRoute = r;
        }
      }

      if (bestId == null || bestRoute == null) break;

      if (fullPath.isNotEmpty && bestRoute.path.isNotEmpty) {
        fullPath.addAll(bestRoute.path.sublist(1));
      } else {
        fullPath.addAll(bestRoute.path);
      }
      fullInstr.addAll(bestRoute.instructions);
      totalDist += bestRoute.distance;
      current = bestId;
      remaining.remove(bestId);
    }

    RouteResult? exitRoute = findShortestPath(current, actualExitId);
    if (exitRoute == null) {
      final exitCandidates = nodes
          .where((n) => _isExitNode(n))
          .map((n) => n.id)
          .toList();
      double bestExitDist = double.infinity;
      for (var eid in exitCandidates) {
        final r = findShortestPath(current, eid);
        if (r != null && r.distance < bestExitDist) {
          bestExitDist = r.distance;
          exitRoute = r;
        }
      }
    }

    if (exitRoute != null) {
      if (fullPath.isNotEmpty) {
        fullPath.addAll(exitRoute.path.sublist(1));
      } else {
        fullPath.addAll(exitRoute.path);
      }
      fullInstr.addAll(exitRoute.instructions);
      totalDist += exitRoute.distance;
    }

    return RouteResult(
      path: fullPath,
      distance: totalDist,
      instructions: fullInstr,
    );
  }

  double _getNodeDist(MapNode n1, MapNode n2) {
    if (n1.latitude != null && n1.longitude != null && n2.latitude != null && n2.longitude != null) {
      return NavigationMath.distanceBetween(n1.latitude!, n1.longitude!, n2.latitude!, n2.longitude!);
    }
    final fp = floorPlans.where((p) => p.id == n1.floorPlanId).firstOrNull ?? (floorPlans.isNotEmpty ? floorPlans.first : null);
    final double scale = (fp != null && fp.scaleMetersPerPx > 0) ? fp.scaleMetersPerPx : 0.035;
    final double dx = n2.x - n1.x;
    final double dy = n2.y - n1.y;
    return max(1.0, sqrt(dx * dx + dy * dy) * scale);
  }

  double _getNodeBearing(MapNode n1, MapNode n2) {
    if (n1.latitude != null && n1.longitude != null && n2.latitude != null && n2.longitude != null) {
      return NavigationMath.bearing(n1.latitude!, n1.longitude!, n2.latitude!, n2.longitude!);
    }
    final double dx = n2.x - n1.x;
    final double dy = n2.y - n1.y;
    return (atan2(dx, -dy) * 180 / pi + 360) % 360;
  }

  List<String> _generateInstructions(List<MapNode> path) {
    if (path.length < 2) return ["You have arrived."];
    List<String> instr = ["Start at ${path[0].name}"];

    for (int i = 0; i < path.length - 1; i++) {
      final cur = path[i], nxt = path[i + 1];
      if (cur.floor != nxt.floor) {
        final connector =
            ['stairs', 'elevator', 'escalator'].contains(nxt.nodeType)
            ? nxt.nodeType
            : 'stairs';
        instr.add("Take the $connector to Floor ${nxt.floor}");
        continue;
      }
      if (i == path.length - 2) {
        final d = _getNodeDist(cur, nxt);
        instr.add(
          "Arrive at ${nxt.name} (${NavigationMath.formatDistance(d)} ahead)",
        );
        continue;
      }
      final nnxt = path[i + 2];
      if (nxt.floor != nnxt.floor) {
        final d = _getNodeDist(cur, nxt);
        instr.add("Walk ${NavigationMath.formatDistance(d)}");
        continue;
      }
      final b1 = _getNodeBearing(cur, nxt);
      final b2 = _getNodeBearing(nxt, nnxt);
      double diff = ((b2 - b1 + 180) % 360) - 180;
      final d = _getNodeDist(cur, nxt);
      if (diff.abs() < 25) {
        instr.add("Walk ${NavigationMath.formatDistance(d)}");
      } else if (diff > 0) {
        instr.add("Turn right at ${nxt.name}");
      } else {
        instr.add("Turn left at ${nxt.name}");
      }
    }
    return instr;
  }
}

class _AdjEntry {
  final String id;
  final double dist;
  final String edgeType;
  _AdjEntry(this.id, this.dist, this.edgeType);
}

class _PQEntry {
  final String id;
  final double dist;
  _PQEntry(this.id, this.dist);
}
